### ISSUE TITLE
implement moduleName when transpiling files

### ISSUE THREAD

**Issue by rolandjitsu** [ivogabe/gulp-typescript#211]
This is mostly for the same reason [#4801](https://github.com/Microsoft/TypeScript/issues/4801) was reported.

As explained in [#3655](https://github.com/Microsoft/TypeScript/issues/3655), when using `system` for the `module` option in `tsconfig.json` (or inline config), the output of the transpilation is:

```js
/// bar.js
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    __.prototype = b.prototype;
    d.prototype = new __();
};System.register(['./foo'], function(exports_1) {
    var foo_1;
    var Bar;
    return {
        setters:[
            function (_foo_1) {
                foo_1 = _foo_1;
            }],
        execute: function() {
            Bar = (function (_super) {
                __extends(Bar, _super);
                function Bar() {
                    _super.apply(this, arguments);
                }
                return Bar;
            })(foo_1.Foo);
            exports_1("Bar", Bar);
        }
    }
});
```
When it should actually be:
```js
/// bar.js
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    __.prototype = b.prototype;
    d.prototype = new __();
};System.register('bar', ['./foo'], function(exports_1) {
    var foo_1;
    var Bar;
    return {
        setters:[
            function (_foo_1) {
                foo_1 = _foo_1;
            }],
        execute: function() {
            Bar = (function (_super) {
                __extends(Bar, _super);
                function Bar() {
                    _super.apply(this, arguments);
                }
                return Bar;
            })(foo_1.Foo);
            exports_1("Bar", Bar);
        }
    }
});
```
Notice the `System.register('bar', ['./foo'], function (exports_1) {});`, it defines the module name. This allows it to be used with SystemJS without having to enable `defaultJSExtensions: true`.

I was about to add the `moduleName?: string` to the [transpile](https://github.com/ivogabe/gulp-typescript/blob/master/lib/compiler.ts#L233) method, but I am not sure how this should work.

Should it detect if `system` is passed as `module` option and if that happens set the `moduleName` to the file name (+ path since it will be required with a path, e.g. `import { Foo } from 'app/foo'` or wherever `foo` resided at when it was compiled).

But then, a source root should also be allowed as option in order to set the `moduleName` to the file path + name absolute to this source root. That is because, let's say that my files are all inside `src/app`, but I do not want to have my module names as `src/app/foo` because the module will be required as `app/foo`. I hope this makes sense.

**Comment by ivogabe** [comment]
I think this should be implemented in TypeScript, not gulp-typescript. The transpile method is only used for isolated modules compilation (file based), whereas most users use the default, project based compilation, so setting the module name there wouldn't solve your issue in all situations.

Besides that, if you're using TS1.6+ the helper functions will be emitted inside the register call, which might solve a part of your issue.

**Comment by rolandjitsu** [comment]
@ivogabe I see, I wasn't aware that `.transpile` is only used for individual file compilation. I am using TS 1.6+, though I am not sure how I could tap into the emitted events using gulp-typescipt, could you elaborate on that a bit?

**Comment by ivogabe** [comment]
> I am using TS 1.6+, though I am not sure how I could tap into the emitted events using gulp-typescript, could you elaborate on that a bit?

What exactly do you want to do? You can add another gulp plugin (after gulp-typescript) that modifies the emitted javascript files.

**Comment by rolandjitsu** [comment]
@ivogabe I know that, but I though you were talking about setting the module name somehow during compilation time.

**Comment by ivogabe** [comment]
No, that's not possible. I meant that changing gulp-typescript internally (so not your configuration), to set the module name option, wouldn't work in all cases.

**Comment by rolandjitsu** [comment]
I see, I figured. Thanks anyway :) I guess I try to use typescript's API instead, that way I can achieve what I want, but I will loose all the benefits of this gulp plugin.

**Comment by frapontillo** [comment]
I am having the same issue, I can't basically use my compiled files at all. @rolandjitsu how did you solve this? As of now SystemJS won't register module-less components at all!

```
Uncaught TypeError: Unexpected anonymous System.register call.
(anonymous function) @ system.src.js:2692
load.metadata.format @ system.src.js:3230 
oldModule @ system.src.js:3725
(anonymous function) @ system.src.js:2423
SystemJSLoader.register @ system.src.js:2648
(anonymous function) @ my-module.js:10
```

If the module name spec can't be added to gulp-typescript, how can this be solved?

**Comment by rolandjitsu** [comment]
@frapontillo how are you compiling and loading? Also, what is your System config? I have it running just fine, the reason I asked was because I wanted to get rid of the `defaultJSExtensions` option, but there are other ways of doing that I guess.

**Comment by frapontillo** [comment]
My compilation task has the following:
```javascript
    var tsResult = gulp
      .src(path.join(conf.paths.src, '**/*.ts'))
      .pipe($.sourcemaps.init())
      .pipe($.typescript(tsProject));

    var dts = tsResult.dts
      .pipe(gulp.dest(path.join(conf.paths.dist, 'definitions')));

    var js = tsResult.js
      .pipe($.concat(conf.fileNames.js))
      .pipe($.sourcemaps.write())
      .pipe(gulp.dest(conf.paths.tmp));

    return $.merge2([dts, js]);
```

and my `tsconfig.json` is:
```json
{
  "compilerOptions": {
    "target": "es5",
    "module": "system",
    "declaration": true
  },
  "sortOutput": true
}
```

The resulting file in `conf.paths.tmp` has all registrations in the following form:
```
System.register([], function(exports_1) {
// ...
});
```
which make SystemJS complain

**Comment by rolandjitsu** [comment]
I see, it's because of your `$.concat(conf.fileNames.js)`. Is there any reason you want to concatenate everything? You could just let it compile to individual files and pipe the files to your dest folder. Then you can config SystemJS:
```js
System.config({defaultJSExtensions: true});
```
in your `index.html` (or where ever you load the concatenated file). Then in the same index file you can load the main script:
```js
System.import('somePath/main').catch(console.log.bind(console));
```
And in your `main.ts` you could import everything else:
```js
import someDefaultExport from './myClass';
import { moreStuff } from './myStuff/moreStuff';
```
Because your concatenation merges all the unnamed `System.register` in one file and System will of course not be able to figure out anything.

One option would be to use browserify which would let you use only one file in your index. I am unsure what other ways there are, but I am pretty sure there must be some way for typescript to compile everything into one file (e.g `tsc --out modules.js`).

**Comment by frapontillo** [comment]
Your take on this is spot on.
In fact, `tsc` (I'm on 1.8.0-dev) is able to accept an `out` file for files compiled with a module loading system, but apparently `gulp-typescript` can't...

I'd like to concatenate everything because I'd like to start using TypeScript in Angular 1.x today, and really don't want to use explicitly use browserify/webpack.
From a "standards" point of view, do you think that JS packages should now be published as separate files or is bundling still the best way to support old projects too?

**Comment by rolandjitsu** [comment]
Glad to help. I actually have a [repo](https://github.com/rolandjitsu/ng-lab) setup of Angular 1 and TS. Perhaps you can get some ideas from it.

The likely reason this gulp plugin does not support that might be because of the TS compiler version you tried not being an official release. Probably it will be implemented, unless it actually already works? @ivogabe any comments on this?

I am not a fan of webpack/browserify either, but it does solve some problems. As for publishing packages, I think that you do want to package/bundle everything in a single file and have sourcemaps as well. It's of course a choice, but it's nice to have a minified and compressed file that users can load and save some bandwidth (since most browsers do not support HTTP/2 yet, so amount of requests matter).

**Comment by frapontillo** [comment]
I tried changing the `tsproject.json` to:
```
{
  "compilerOptions": {
    "target": "es5",
    "module": "system",
    "declaration": true,
    "out": "out.js"
  },
  "sortOutput": true
}
```
but it still outputs single files and an `out.js` file with only sourcemaps in it.

I'd love if @ivogabe could chime in.

**Comment by ivogabe** [comment]
> The likely reason this gulp plugin does not support that might be because of the TS compiler version you tried not being an official release. Probably it will be implemented, unless it actually already works? @ivogabe any comments on this?

You can use a custom version of TypeScript, thus a nightly build or even a fork. See the readme for details. The `out` option should be working then. Can you get it working this way?

I must say I haven't used SystemJS myself (so I'm glad @rolandjitsu can help here).

@frapontillo With `out` you won't need the `sortOutput` option, and you should provide it in the `compilerOptions` section, instead of the root.

**Comment by frapontillo** [comment]
Specifying typescript in the options made it work! I now have a single file with such registrations:
```
System.register("src/path/to/file", [], function(exports_1) {
// ...
});
```

Is there any way to get rid of the `src/` prefix? My guess is that it's there because `tsconfig.json` is in the root directory.

**Comment by rolandjitsu** [comment]
I think there is a `rootDir` (or something like that, check [tsconfig schema](http://json.schemastore.org/tsconfig)) option that might strip that. Set it to either `./` or `./src`, which ever suits your needs. But I am not sure it will do what I think it does, perhaps combine with `outDir`.
